tool/tbot/init.go

```diff
@@ -0,0 +1,534 @@
+		}
+
+		if destination == nil {
+			// TODO: in the future if/when other backends are supported,
+			// destination might be nil because the user tried to enter a non
+			// filesystem path, so this error message could be misleading.
+			return trace.NotFound("Cannot initialize destination %q because "+
+				"it has not been configured.", cf.InitDir)
+		}
+	}
+
+	destImpl, err := destination.GetDestination()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	destDir, ok := destImpl.(*config.DestinationDirectory)
+	if !ok {
+		return trace.BadParameter("`tbot init` only supports directory destinations")
+	}
+
+	log.Infof("Initializing destination: %s", destImpl)
+
+	// Create the directory if needed. We haven't checked directory ownership,
+	// but it will fail when the ACLs are created if anything is misconfigured.
+	if err := destDir.Init(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Next, test if we want + have ACL support, and set aclOpts if we do.
+	// Desired owner is derived from the ACL mode.
+	var aclOpts *botfs.ACLOptions
+	var ownerUser *user.User
+	var ownerGroup *user.Group
+
+	switch destDir.ACLs {
+	case botfs.ACLRequired, botfs.ACLTry:
+		log.Debug("Testing for ACL support...")
+
+		// Awkward control flow here, but we want these to fail together.
+		ownerUser, ownerGroup, aclOpts, err = getAndTestACLOptions(cf, destDir.Path)
+		if err != nil {
+			if destDir.ACLs == botfs.ACLRequired {
+				// ACLs were specifically requested (vs "try" mode), so fail.
+				return trace.Wrap(err, aclTestFailedMessage, destImpl)
+			}
+
+			// Otherwise, fall back to no ACL with a warning.
+			log.WithError(err).Warnf(aclTestFailedMessage, destImpl)
+			aclOpts = nil
+
+			// We'll also need to re-fetch the owner as the defaults are
+			// different in the fallback case.
+			ownerUser, ownerGroup, err = getOwner(cf.Owner, "")
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		} else if aclOpts.ReaderUser.Uid == ownerUser.Uid {
+			log.Warnf("The destination owner (%s) and reader (%s) are the "+
+				"same. This will break OpenSSH.", aclOpts.ReaderUser.Username,
+				ownerUser.Username)
+		}
+	default:
+		log.Info("ACLs disabled for this destination.")
+		ownerUser, ownerGroup, err = getOwner(cf.Owner, "")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Next, resolve what we want and what we already have.
+	desired, err := getInitArtifacts(destination)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	existing, err := getExistingArtifacts(destDir.Path)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	toCreate := diffArtifacts(desired, existing)
+	toRemove := diffArtifacts(existing, desired)
+
+	// Based on this, create any new files.
+	if len(toCreate) > 0 {
+		log.Infof("To create: %v", toCreate)
+
+		for key, isDir := range toCreate {
+			path := filepath.Join(destDir.Path, key)
+			if err := botfs.Create(path, isDir, destDir.Symlinks); err != nil {
+				return trace.Wrap(err)
+			}
```

Comment:
If this fails should we clean up any files already created?

Reply:
It's idempotent, so I think the workflow for failure here is to correct the issue and re-run anyway. Cleaning up in general is a little bit scary (we only do it with `--clean` right now) and likely to fail anyway if a create just failed.